lib/internal/streams/from.js

```diff
@@ -34,21 +34,48 @@ function from(Readable, iterable, opts) {
     objectMode: true,
     ...opts
   });
+
   // Reading boolean to protect against _read
   // being called before last iteration completion.
   let reading = false;
+
+  // needToClose boolean if iterator needs to be explicitly closed
+  let needToClose = false;
+
   readable._read = function() {
     if (!reading) {
       reading = true;
       next();
     }
   };
+
+  readable._destroy = async function(error, cb) {
+    try {
+      if (needToClose) {
+        needToClose = false;
+        if (typeof iterator.return === 'function') {
+          const { value } = await iterator.return();
+          await value;
+        }
+      }
+    } catch (e) {
+      if (!error) { // Do not hide current error in case if .return() throws
+        error = e;
+      }
+    } finally {
+      cb(error);
+    }
+  };
+
   async function next() {
     try {
+      needToClose = false;
       const { value, done } = await iterator.next();
+      needToClose = !done;
+      const resolved = await value;
```

Comment:
Need to close the iterator if `readable.destroyed` after `await`.

Reply:
fixed.
Updated tests to catch that.